### ISSUE TITLE
Missing prop validation in React.FunctionComponent

### ISSUE THREAD

**Issue by dchebakov** [yannickcr/eslint-plugin-react#2353]
**Tell us about your environment**

- **ESLint Version**: v6.0.1
- **Node Version**: v10.16.0
- **npm Version**: v6.9.0

**Configuration**
```javascript
module.exports = {
  parser: '@typescript-eslint/parser',
  parserOptions: {
    project: './tsconfig.json',
  },
  plugins: ['@typescript-eslint'],
  extends: [
    'plugin:react/recommended',
    'plugin:@typescript-eslint/recommended',
  ],
}
```
**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**
```typescript
import * as React from 'react';
interface PersonProps {
    username: string;
}
const Person: React.FunctionComponent<PersonProps> = (props): React.ReactElement => (
    <div>{props.username}</div>
);
```

```bash
npx eslint src --ext .tsx
```

**What did you expect to happen?**
`eslint` show no errors, because type definition actually present in generic type `React.FunctionComponent<PersonProps>`

**What actually happened? Please include the actual, raw output from ESLint.**
```bash
error    'username' is missing in props validation       react/prop-types
```

**Comment by ljharb** [comment]
Technically you typed the variable and not the function; what happens if you use a normal non-arrow function for your component?

**Comment by dchebakov** [comment]
> Technically you typed the variable and not the function; what happens if you use a normal non-arrow function for your component?

For non-arror function component:
```typescript
function Person(props: PersonProps): React.ReactElement {
    return (
        <div>{props.username}</div>
    )
}
```
there are no errors, but I want to use arrow func with generic type FunctionComponent. For that case, type of props should be inferred from `React.FunctionComponent<PersonProps>`

**Comment by ljharb** [comment]
That’s something typescript does, but unless the typescript eslint parser passes those types to this plugin, we have no way of leveraging them.

It’s a better practice to use non arrows for SFCs anyways, so they have a proper reliable name for debugging.

**Comment by dchebakov** [comment]
Should I create an issue for typescript parser?

**Comment by ljharb** [comment]
I’m not sure. You could certainly try a PR here to improve the type detection - before filing an issue even on the typescript parser you’d need to have tested and figured out what was needed.

The simplest fix, though, is to follow what the most popular styleguide recommends and use normal functions for components.

**Comment by tx-steven** [comment]
I'm having this same error but with functional components. Can't seem to find any good reason why it's happening. I have a functional component. With a typescript props interface. When I access a property off of the prop. I get `<property> is missing in types validation`.

**Comment by ljharb** [comment]
@tx-steven is it an arrow function or a normal function?

**Comment by skube** [comment]
Adding `: PersonProps` somewhat redundantly seems to work...

```tsx
const Person: React.FunctionComponent<PersonProps> = (props: PersonProps): React.ReactElement => (
    <div>{props.username}</div>
);
```

**Comment by ljharb** [comment]
That’s not redundant; in your case the first one is typing the variable, not the function.

**Comment by skube** [comment]
I guess but I thought TypeScript [infers the types](https://www.typescriptlang.org/docs/handbook/functions.html#inferring-the-types) — suggesting that it's not required to do both left and right operands.

**Comment by shulcsm** [comment]
It does, this rule is redundant for arrowfunc: React.FC<> case since props are checked by TS

**Comment by ljharb** [comment]
TypeScript may, but if the eslint typescript parser doesn’t, there’s nothing this plugin can do.

**Comment by Rey-Wang** [comment]
@skube it works

**Comment by aaronlna** [comment]
@skube A small formality, but the inferred type is more specifically `React.PropsWithChildren<PersonProps>`. Although if you aren't using any of its (PropsWithChildren) attributes, `PersonProps` would be sufficient.

If this isn't an eslint issue would creating an option to ignore this case be worthwhile?

**Comment by kristojorg** [comment]
This is quite annoying, and I don't believe there is any value in using non arrow functions for function components.

**Comment by ljharb** [comment]
it has lots of value; explicitly that you aren’t relying on name inference, which is unintuitive.

**Comment by mohsinulhaq** [comment]
> It’s a better practice to use non arrows for SFCs anyways, so they have a proper reliable name for debugging.

One advantage of using arrow function is the `children` typing you get for free. For example:

```
interface CompProps {
    prop1: number
}

const Comp: React.FC<CompProps> = ({ children, prop1 }) => {...} // children inferred from React.FC

function Comp({ prop1 }: CompProps) {...} // where to put React.FC?
```
 ref: [TypeScript#16918](https://github.com/microsoft/TypeScript/issues/16918)

It works in TSLint but not in ESLint.

**Comment by skube** [comment]
> function Comp({ prop1 }: CompProps) {...} // where to put React.FC?

Can't you do? :
```tsx
function Comp({ prop1 }: CompProps): React.ReactElement {...} 
```

**Comment by mohsinulhaq** [comment]
> > function Comp({ prop1 }: CompProps) {...} // where to put React.FC?
> 
> Can't you do? :
> 
> ```tsx
> function Comp({ prop1 }: CompProps): React.ReactElement {...} 
> ```

Sorry I missed destructuring children in the second component. Added it now. After that, you face the same problem.

**Comment by xjamundx** [comment]
I found this to be an interesting case

```ts
// fails with prop-types
export const BlockBad: React.FC<BlockProps> = ({ attributeName }) => {
  return <div data-attributeName={attributeName} />;
};

// does not fail
export const BlockGood: React.FC<BlockProps> = ({ attributeName }) => {
  if (attributeName) {
    return <div data-attributeName={attributeName} />;
  }
};
```

For those seeking ideas about how to type these things. I found both of these resources helpful:
- https://github.com/typescript-cheatsheets/react-typescript-cheatsheet#function-components
- https://github.com/piotrwitek/react-redux-typescript-guide/blob/master/README.md#function-components---fc

Basically one suggests to use `function` and the other suggests the approach being used by a lot of people in this thread `const val: React.Fc = () => {}`

**Comment by jawang94** [comment]
Not an ideal world solution but this should resolve your ESLint error. `React.FC<Props>` satisfies TS and `({ username }: <Props>)` satisfies ESLint.

```
type Props = {
    username: string;
};

const Person: React.FC<Props> = ({ username }: <Props>) => (
    <div>{username}</div>
);
```

**Comment by leonardorb** [comment]
@xjamundx very nice. This will work without warnings with the latest versions:

```tsx
import React from 'react'

type CellProps = {
  value: string
}

const Cell: React.FunctionComponent<CellProps> = ({ value }: CellProps) => {
  return <div className="item">{value}</div>
}

export default Cell
```

**Comment by saranshkataria** [comment]
@leebyron that will work, but as @xjamundx pointed out, you are typing it twice. It should have been inferred automatically instead.

Another potential way of doing it would be: 
const Cell = ({ value }: CellProps): JSX.Element => {
  return <div className="item">{value}</div>
}

But not sure what the ideal statement should be.

**Comment by AlaaZorkane** [comment]
Any updates on this? Or at least a best practice, I don't see how is typing twice is acceptable in this case...

**Comment by saranshkataria** [comment]
I have manually turned off the react/prop-types rule in my config for now. Not sure if there is a better way to handle it.

**Comment by tx-steven** [comment]
Best practice is to just not use arrow functions so I don't think you guys are going to find an easy/simple work around for this if you want to use arrow functions for react components

**Comment by TeoTN** [comment]
I think the rule is pretty useless in the current shape - it's supposed to check if props are typed, not to enforce whether the project should use arrow functions or not

**Comment by ljharb** [comment]
@TeoTN unfortunately by using arrow syntax it seems TS eslint parsers are incapable of attaching the type to an arrow function, so here we are :-/ happy to review a PR that can improve the situation.

**Comment by TeoTN** [comment]
I don't have a solution at hand. If it's unfeasible at the moment, it should be probably removed from the default set of rules.

**Comment by ljharb** [comment]
What is "it" here? There's a number of reasons why components shouldn't be arrow functions; it's totally fine if that truth is revealed by various rules.

**Comment by abmantis** [comment]
@ljharb Why do you say components shouldn't be arrow functions? I haven't found the reasons yet (not saying they don't exist).

**Comment by ljharb** [comment]
@abmantis arrow functions don't have explicit names, only sometimes-inferred ones. An explicit name is critical for debuggability and greppability.

**Comment by TeoTN** [comment]
Arrow functions don't hold names by default, but for debugging purposes they can be assigned a `displayName` field, which enables finding that name in React Profiler, if you need it.
"Greppability" is not a great argument either, you can grep for the name of the variable holding it, also any IDE will find usages of arrow function by its variable.

My key point here is that `react/prop-types` rule that is by default enabled in the preset, is supposed to check if prop types are provided, not someone is using arrow functions. Enforcing debatable code style is not within the scope of the rule intent, is it?
If the rule is not fulfilling its purpose, because it doesn't work with one of the most common ways of writing components, it shouldn't be a part of recommended linting rules for React. Probably we shouldn't enforce code style, there's prettier for that, there should be only genuine issues detected.

**Comment by ljharb** [comment]
You're right about `displayName` - you can also `Object.defineProperty` a `name` on it just fine, but the point is that it's extra effort most people fail to do.

It's supposed to ensure types are provided. It works fine on propTypes for arrow functions - it's just a failing in the TS parser that it can't properly type arrow functions.

Use propTypes, and you can use arrow functions all you want. The rule is working perfectly.

**Comment by abmantis** [comment]
@ljharb but if I have `const MyCompo = (props) => { ... }` won't MyCompo be inferred for the component and show in the debugger as if it was a traditional function, without the need for `displayName`?

**Comment by ljharb** [comment]
@abmantis yes, but function name inference is implicit, doesn't work in every browser engine you might be debugging in (or collecting stack traces from), and I'm saying that for clarity/explicitness, it should never be relied upon.

**Comment by TeoTN** [comment]
> It's supposed to ensure types are provided. It works fine on propTypes for arrow functions - it's just a failing in the TS parser that it can't properly type arrow functions.

The problem is that it doesn't always work, and because of that it enforces debatable code style, instead of fulfilling its main purpose. I think that arrow functions are better, because they save me a few bytes of the name in the bundle, and that's more important than having debug names. I can always add them ad-hoc when I need. 
This example is not a genuine style problem but, unfortunately, your personal preference, that is no better than anyone's  preference.

**Comment by ljharb** [comment]
The bundle gets gzipped; you save nothing consequential, and that is very much less important than debuggability anyways.

**Comment by TeoTN** [comment]
I think you're consistently missing my point, that the purpose of this rule is to check props - it is not `react/enforce-function-keyword`. It's cool that you want to always write the lengthy way and you like it but its not the only way to ensure you have debug names, and there's no reason to push your solution. But the key point here is that the rule is not serving its purpose well not that it has an unrelated side-effect that is opinionated and probably unneeded.

**Comment by ljharb** [comment]
I agree that due to a flaw in the typescript eslint parser, *when using typescript* and when writing components as arrow functions, you may find this rule annoying. If that's the case, feel free to turn it off - but the rule is functioning perfectly, the side effect doesn't affect most people since non-arrow components are a best practice and the majority aren't using TS, and I encourage you to file an issue on the TS eslint parser if you want it addressed.

Since there's nothing actionable for this project here until the TS eslint parser is fixed, I'm going to close this, but will be more than happy to reopen if that changes.

**Comment by GuiCavi** [comment]
I know this is closed until new changes, just want to drop here an info about something I didn't read in above comments.

```ts
interface Props {
   prop1: string;
}

// If you do this alternative, you can't extract children from props.
// This means you have to put children in Props interface always.
const CustomComponent: React.FC<Props> = ({ prop1 }: Props) => (...)
```

Same for this alternative
```ts
interface Props {
   prop1: string;
}

const CustomComponent = ({ prop1 }: Props) => (...)
```

**Comment by ljharb** [comment]
You should have to have `children` in your propTypes / props types anyways, it's a prop like any other.

**Comment by GuiCavi** [comment]
Oh, I just realize it's missing one piece of explanation, sorry for that. According to React:

```ts
type PropsWithChildren<P> = P & { children?: ReactNode };

/// ...

interface FunctionComponent<P = {}> {
    (props: PropsWithChildren<P>, context?: any): ReactElement<any, any> | null;
    propTypes?: WeakValidationMap<P>;
    contextTypes?: ValidationMap<any>;
    defaultProps?: Partial<P>;
    displayName?: string;
}
```

all FunctionalComponents must have a `children` prop when using typing like `const t: React.FC<T>`.

But hey, I'm not disagreeing with you, just here to add info to the conversation.

**Comment by shirly-chen-awx** [comment]
"overrides": [
        {
            "files": ["**/*.tsx"],
            "rules": {
                "react/prop-types": "off"
            }
        }
    ]

**Comment by osorina** [comment]
If someone still faces this problem - update your dependencies

**Comment by icfantv** [comment]
Just a head's up on @shirly-chen-awx's post.  Not sure if it's because I'm using `@typescript-eslint` or having WebStorm live running the linting,  but `off` is not a valid value.  I had to use the numeric `0` (as opposed to `"0"`) to turn it off.

**Comment by ljharb** [comment]
@icfantv what version of eslint? The two valid values to disable a rule have been `"off"` and `0` since at least v3.

**Comment by icfantv** [comment]
@ljharb latest version, 7.16.  But now I can't reproduce the error (and don't remember what I did to get it to fire...).  🤷   Could be that it was one last dig by 2020 and now that it's 2021, everything is awesome again.  Ignore my original comment...

**Comment by ITenthusiasm** [comment]
> It’s a better practice to use non arrows for SFCs anyways

@ljharb I typically prefer to write non arrows anyway, but is there any chance you could expound on why this is the case? (I noticed you mentioned the `displayName`, which is another reason I prefer regular functions. But that's technically just one reason.)

More specifically, I'm noticing that whereas with regular JS I can do:

```jsx
function MyComponent(props) {
  // ...
}

MyComponent.propTypes = {
  // ...
};

MyComponent.defaultProps = {
  // ...
};
```

It doesn't seem to be as easy to do this in TypeScript:

```typescript
interface PropTypes {
  // ...
}

function MyComponent(props: PropTypes) {
  // ...
}

MyComponent.defaultProps = {
  // ...
};
```

More specifically, I seem to get an error saying that `defaultProps` doesn't exist on the functional component. I'm aware that typically destructuring is the preference to get around this issue. And it's a commonly encouraged practice as well. But I imagine there are use cases where someone would just want to use `props` and then access the properties accordingly in their code (tbf it's just an assumption). Such a situation necessitates using `MyComponent.defaultProps`, and seeing as it's doable in regular JS it would make sense for it to be doable in TS too I think.

I know `React.FC<PropTypes>` circumvents this issue. It does tell TypeScript to expect a `defaultProps` property. Are you aware of a similar solution with regular functions?

Thanks.

**Comment by ljharb** [comment]
@ITenthusiasm the explicit name is a pretty big reason - arrows can only ever have inferred (or defineProperty'd names), and `.displayName` is just for react, not for actual engine debugging.

TS surely has issues doing many normal JS things. I assume there's a way to use normal functions and the `React.FC<PropTypes>` type.

**Comment by ITenthusiasm** [comment]
Fair enough. Thanks!

**Comment by gearFifth** [comment]
> Adding `: PersonProps` (somewhat redundantly) seems to work...
> 
> ```tsx
> const Person: React.FunctionComponent<PersonProps> = (props: PersonProps): React.ReactElement => (
>     <div>{props.username}</div>
> );
> ```

with this you lose the default props that come with the functional component like the children prop.

**Comment by gearFifth** [comment]
> I know this is closed until new changes, just want to drop here an info about something I didn't read in above comments.
> 
> ```ts
> interface Props {
>    prop1: string;
> }
> 
> // If you do this alternative, you can't extract children from props.
> // This means you have to put children in Props interface always.
> const CustomComponent: React.FC<Props> = ({ prop1 }: Props) => (...)
> ```
> 
> Same for this alternative
> 
> ```ts
> interface Props {
>    prop1: string;
> }
> 
> const CustomComponent = ({ prop1 }: Props) => (...)
> ```

with this you lose the default props that come with the functional component like the children prop.

**Comment by ITenthusiasm** [comment]
Just for documentation purposes, we hashed some of this out on typescript-eslint/typescript-eslint#3075. There seems to be a solution if someone is willing to open up a PR.

I'm not as familiar with making code changes for ESLint, and my hands have been a little tied as of recent. If no one picks it up, maybe I can take a stab at it in the future? But in the meantime, if anyone is looking for a solution to come sooner, I'd refer to the linked issue.

**Comment by ITenthusiasm** [comment]
If it's any additional help while traversing the comments, then with respect to this one:

> Common cases it could handle:
> - React.SFC
> - React.StatelessComponent
> - React.FunctionComponent
> - React.FC

Recall that `React.SFC` and `React.StatelessComponent` have been deprecated. And `React.FC` is just shorthand for `React.FunctionComponent`. So as you're bouncing through the cases you want to handle...you really only need to be concerned about `React.FunctionComponent`.

**Comment by rawandeheliah** [comment]
I just stopped tracking this issue :
by adding ` "rules": {
                "react/prop-types": "off"
            }`
in .eslintrc.js

**Comment by ITenthusiasm** [comment]
Returning to this conversation again after quite some time to make 2 comments. The first one is just to get people thinking. You don't have to care about it. 🙂 

First, although I mentioned earlier that I typically use regular functions, I have found `React.FC` appealing in the past because it created a simple way to make components in React with TypeScript. The place I'm working at now uses `React.FC` frequently. However, after trying to make some re-usable and non-simple components, I realized something and wanted to give a warning:

I think we're used to `React.FC` because it's what we're given out of the box. But actually, `React.FC` makes your components less specific than they should be. Here are 2 basic examples. There are likely more.

## 1) Children

From [gearFifth's comment](https://github.com/yannickcr/eslint-plugin-react/issues/2353#issuecomment-808179060) ([mohsinulhaq's comment](https://github.com/yannickcr/eslint-plugin-react/issues/2353#issuecomment-575933942) had something similar):
> with this you lose the default props that come with the functional component like the children prop.

At first this made me sad. But I've learned this is actually a _really_ good thing. Although in reality all React components _can_ take a `children` prop, not all components are _intended_ to take a `children` prop. Similarly, any regular function you define _can_ take more than the specified number of arguments. However, that's not the intent of the function and it won't accomplish anything. Supplying extraneous arguments will only cause confusion. This is part of the point of TS to begin with.

"But I actually intend to use the `children` prop for such and such component". That's a valid argument. But in that case, it's a much better developer experience to explicitly supply `children` in your `ComponentProps` type/interface for two reasons:

1) It tells the developer that this component _explicitly_ requires or intends to use the `children` prop.
2) It tells the developer what is _allowed_ for the `children` prop, preventing any confusion.

On that second note, let's say that you have a `Card` component that's intended to take text as its children. And it _only_ expects text. The default `children` prop from `React.FC` is a `ReactNode`. This means TS will not care if your developer passes some jank JSX into the component and breaks it because you didn't tell them what was expected by your types. Thus, it's friendlier to explicitly type your `children` anyway.

Unfortunately, `React.FC` also has the potential to get in the way of your explicitly defined `children` prop. For instance, when I try to say "Hey, use a `string` child", `React.FC` says, "Oh. You wanted `string & ReactNode` for `children`". In the end, TS seems to restrict it properly. But that can still potentially be a source of confusion.

## 2) Return Types

This is the thing that really bothers me, to be honest. I absolutely hate specifying return types on components. And yet... I've come to see they're useful while making some re-usable or non-simple components. The default return type from `React.FC` is nice, but it's also not always valid. `React.FC` leaves the door open for a component to return `null`, but maybe you have components that are never intended to return `null`. Again, stricter, more explicit type guards protect you from this issue. They also prevent TypeScript from complaining about how a component could be `null` when you know you've written your code so that it never would be `null`.

--- 

Regarding this

```typescript
type PropsWithChildren<P> = P & { children?: ReactNode };

/// ...

interface FunctionComponent<P = {}> {
    (props: PropsWithChildren<P>, context?: any): ReactElement<any, any> | null;
    propTypes?: WeakValidationMap<P>;
    contextTypes?: ValidationMap<any>;
    defaultProps?: Partial<P>;
    displayName?: string;
}
```

`defaultProps` and `propTypes` are taken care of by TS. `contextTypes` and the `context` argument are [legacy](https://reactjs.org/docs/legacy-context.html) from what I understand and are taken care of by [useContext](https://reactjs.org/docs/hooks-reference.html#usecontext). And an explicit `children` prop and return type provide a better developer experience with clearer type specificity (..._long term_, that is... I get the pain of not liking to write it out).

Of course, you may not always care about this level of specificity. But oftentimes you do, and that's worth nothing. (And if you don't care, it might be worth asking why you're interested in TS.) Defaulting to regular functions instead of `React.FC`s increasingly seems like more of a good idea to me as I learn more (except for the fact that I have to keep specifying return types :pensive:).

The problem with the default `FC` type provided by React is that it tries to be JavaScript and TypeScript at the same time. That is, it acknowledges the fact that the JS functionality will allow you to do whatever you want (like adding extraneous args). And it seems to try to keep the type "backwards compatible" for anyone working with legacy code. So it creates a TS type to cover all of the use cases. But again, this defeats the point of the safety TS aims to provide to begin with. Instead, they should've provided the bare minimum types for a functional component.

Stated differently (and perhaps far too strongly), `React.FC` as it is now -- in _some_ sense -- is the React version of using `any` in TS... but for functional components.

Consider these things when you write your own components...and in the significance of this overall discussion.

**Comment by ITenthusiasm** [comment]
Second, despite everything that I said in that first comment of mine above, I think there is still an issue that needs to be addressed.

We have an ESLint rule that can't be disabled by default because it provides good common standards to follow. (There are other React developers that don't use TS. :smirk:) At the same time, the linting does seem to be incorrect here for TS. In a theoretical world where `React.FC` was improved to have the bear minimum types, I'm assuming that we'd still run into this issue. So if possible, this problem warrants fixing.

The comments about turning this rule off locally are becoming redundant, and they aren't providing a true solution to the problem. At the same time, comments about only using non-arrow functions still leave the problem at hand. If it's truly important, restricting components to regular functions would need to be its own ESLint rule...though idk how it would be written. But that shouldn't bleed into this discussion.

Although I haven't been able to do anything with it myself yet, the discussion at typescript-eslint/typescript-eslint#3075 provides guidance towards a solution to this problem if I understood it correctly. Since this issue is important, if it _is_ actionable, then @ljharb could you reopen this issue? If it's nothing that's being immediately prioritized, maybe something like a `help wanted` tag would be fine. But an open issue would more clearly specify intent here.

**Comment by ljharb** [comment]
@ITenthusiasm i'd prefer a new issue be filed that clearly outlines the problem and the needed solution - this thread is quite long and full of distractions.

Yes, there ideally should be an eslint rule forbidding functional components to be written as arrow functions; i'm not sure why we don't have one yet.

**Comment by ITenthusiasm** [comment]
Thanks for the fast response. I'll look into adding a new issue soon if no one else does.

**Comment by jiminikiz** [comment]
> @ITenthusiasm i'd prefer a new issue be filed that clearly outlines the problem and the needed solution - this thread is quite long and full of distractions.
> 
> Yes, there ideally should be an eslint rule forbidding functional components to be written as arrow functions; i'm not sure why we don't have one yet.

GUYS GUYS GUYS!

@ljharb Please forgive me for not filing a new issue, but I do believe I have a REAL solution to this problem. Requires no change to `eslint`.

![Screen Shot 2021-07-07 at 3 33 43 PM (2)](https://user-images.githubusercontent.com/709908/124831912-1f86f400-df39-11eb-9b2f-605a472ae1a7.png)

Turns out `FC` takes `type` as its parameter, **NOT** `interface`

This works => WITH NO LINT ERRORS 😎  (via `eslint`):

```jsx
import { FC, PropsWithChildren } from 'react';

export type PersonProps = {
  name: string;
  age: number;
  email: string;
};

export const PersonCard: FC<PropsWithChildren<PersonProps>> = (Props) => {
  const { children, name, age, email } = Props;

  return (
    <div>
      <ul>
        <li>name: {name}</li>
        <li>age: {age}</li>
        <li>email: {email}</li>
      </ul>
      {children}
    </div>
  );
};

export default PersonCard;
```

**Comment by rsimp** [comment]
My two cents:
1. While you can make a rule for enforcing arrow functions, realize they are necessary for React.memo, React.forwardRef, or any other type of composition function. Also while they are better for automatically setting display name, any project that uses babel will still have this done by simply using the preset for react. You can also handle setting displayName through something like babel plugin macro. IMO this should never be part of the recommended ruleset, it's too dependent on the toolchain to decide whether it's a good policy or not. Considering React.memo forces you to use variable assignment (blowing away the main reason here for using `function`) there's even an argument for always using fat arrows for consistency.
2. React.FC was removed from the [create-react-app typescript template](https://github.com/facebook/create-react-app/pull/8177?expand=1) because of how it handles `children`. While a convenient shorthand, you can no longer make `children` required or disallowed, or restrict it further than ReactNode which is pretty broad. If you're fine with the looser typings on children it's still pretty convenient. I don't think it's relevant to this discussion but I think the community is moving away from it.
3. You can always turn off this rule and just use explicit module boundary types from typescript-eslint

**Comment by ljharb** [comment]
@rsimp Arrow functions are in no way necessary for any kind of component function, including memo and forwardRef, and arrow functions are WORSE for setting the display name, and the react preset won’t always ensure it’s present. React.memo takes a function as an argument, so variable assignment can’t possibly infer the display name.

**Comment by rsimp** [comment]
> @rsimp Arrow functions are in no way necessary for any kind of component function, including memo and forwardRef, and arrow functions are WORSE for setting the display name, and the react preset won’t always ensure it’s present. React.memo takes a function as an argument, so variable assignment can’t possibly infer the display name.

I may have the wrong babel preset listed but here's a quote from [gaearon in a react-devtools issue](https://github.com/facebook/react-devtools/issues/267#issuecomment-241042995)

> If you use Babel, es2015 preset contains function-name plugin that infers the name based on left hand side in expressions.
> 
> For Facebook internal use case, @spicyj recently fixed it by enabling the same transform.

As to `React.memo` I believe only in React 17 and only since late 2020/early 2021 can you get a nice `displayName` with a named function without setting it explicitly. If `eslint-plugin-react` wanted to enforce named functions you'd want to include inlined named functions inside of `React.memo` too. Also unless you used a default export you'd be naming the component twice in your code. Once for the named function, and again for the `React.memo` variable assignment. Or you could use a babel macro to prevent the duplication and set it for you.

My main point is that tooling can be setup so that a component created with a fat arrow function can have a display name without you explicitly setting it. If someone (and I don't think it's that uncommon) has this kind of project setup, then does it make sense to put an "enforce function components" rule into the recommended ruleset? I don't think it's a bad idea for a rule, but you don't want a rule to be "recommended" if it'll collide with projects on something this arbitrary. All that should matter is that `displayName` is set. `eslint-plugin-react` shouldn't be enforcing *how* it's done. If there's not enough information to do it automatically then you shouldn't be setting the rule automatically either

**Comment by rsimp** [comment]
@ljharb you're right though, fat arrow is definitely not necessary for `React.memo` `React.forwardRef`, thanks for the correction. But I think the main point still stands. In a lot of cases with function composition, display name resolution often gets tricky. Simply enforcing named functions isn't enough. You'd need to actually read each library's code, use a code mod, or just start setting it yourself.

Named functions would necessary for the library to be capable of setting it automatically, but won't guarantee that they will. It also get's pretty verbose using more than one composition function together. Let's say `React.memo` and `React.forwardRef`. I'm sure people will want to handle these types of cases in different ways. I'm not sure it makes for a good "recommended" rule. Just my opinion.

**Comment by ljharb** [comment]
For React, named functions are absolutely enough. If you can find a popular tool that doesn't respect a component's name when available, call it out and i'll send them a PR fixing it.

Dan himself says that components shouldn't be arrow functions, because of the name in particular. It is a perfect candidate for being a recommended rule - components should be regular functions.

**Comment by Chaoste** [comment]
I would like to mention another case where I couldn't resolve the linting errors after testing all the proposed solutions in this thread (turning off this lint rule should not be the solution but rather a temporary workaround):

```jsx
import React from 'react';

type MySvgIconProps = React.SVGProps<SVGSVGElement>;

export function MySvgIcon(props: MySvgIconProps) {
  const withDefaults = {
    ...props,
    width: props.width ?? 100,
    height: props.height ?? 100,
    color: props.color ?? '#fff',
  };

  return (
    <svg
      xmlns="http://www.w3.org/2000/svg"
      enableBackground="new 0 0 24 24"
      viewBox="0 0 24 24"
      {...withDefaults}>
      <rect fill="none" height="24" width="24" />
      <path
        d="M1,1C20,20,16,-8"
        fill={withDefaults.color}
      />
    </svg>
  );
}
```

Error:
```bash
   9:18  error  'width' is missing in props validation   react/prop-types
  10:19  error  'height' is missing in props validation  react/prop-types
  11:18  error  'color' is missing in props validation   react/prop-types
```

Modules:
```
"@typescript-eslint/eslint-plugin": "3.10.1",
"@typescript-eslint/parser": "3.10.1",
"eslint": "7.26.0",
"@types/react": "16.14.6",
"react": "16.14.0",
```

**Comment by colemars** [comment]
This rule is just doubling up on the job TS is already doing. Don't recommend.

**Comment by ljharb** [comment]
@colemars no, it's not, because it can check runtime values, and subsets like integers, and all sorts of things TS is capable of doing. In addition, TS has the flaw that it doesn't properly differentiate the external types of your props, with the internal type of your props, so you can run into situations where your types lie to you.

Always use propTypes, *especially* when using TS.

**Comment by soujvnunes** [comment]
I think I'm late here, although if you are using ESlint with TypeScript, be sure to extend `"plugin:@typescript-eslint/recommended",` in the `.eslintrc` project file.

**Comment by AdrianDiazG** [comment]
why is this thread closed? the way this rule works right now seems obviously redundant if you wanna use arrow functions and React.FC, and no, changing to the other function definitions doesn't look like a good solution, it's just ignoring the problem

**Comment by ljharb** [comment]
@AdrianDiazG you're not supposed to do that tho - components should be named functions.

**Comment by jiminikiz** [comment]
> Meanwhile YEARS LATER ...

SO!

If you disable this rule:

```ts
{
  // .eslintrc (or equivalent)
  "rules": {
     "react/function-component-definition": "off",
  }
}
```

TypeScript will take care of the rest and the following pass type checking without having to import `PropsWithChildren`:

```ts
import { FC } from 'react';
export type Item = HTMLAttributes<HTMLDivElement> & {
  requiredProp: string;
  optionalProp?: number
}

export const Item: FC<StackItemProps> = ({
  children,
  prop1,
  prop2 = 2,
  ...props,
}) => {
  console.debug({ prop1, prop2 });
  return (
    <div {...props}>
      {children}
    </div>
  );
};

export default Item;
```

I hope this helps someone!

**Comment by ljharb** [comment]
Double-typing props is ideal, though, because it provides both build time and runtime validation.

**Comment by jiminikiz** [comment]
@ljharb At least as I understand TS, double typing the `props` being passed within a `function` definition does _not_ provide any extra runtime validation. In fact, `typing` in general does not provide any runtime validation as all of that goes away once compiled down to ES5... right?

This is a great read, it specifically goes over some runtime type checking strategies:
https://blog.logrocket.com/methods-for-typescript-runtime-type-checking/

Not trying to be right here. Just trying to get it right.

**Comment by ljharb** [comment]
@jiminikiz what i mean is, using propTypes *and* types is valuable. Using double type syntax, obviously, isn't valuable.

Either way, the impact of this rule affects *observable runtime properties* of the component, which is why it's important it not be an arrow function, and causing negative runtime impact because of type system concerns seems like a wildly inverted set of priorities.

**Comment by jiminikiz** [comment]
> @jiminikiz what i mean is, using propTypes and types is valuable. 

Go for it. I am not here to talk about `.propTypes`. I hear your concerns on that.

The devs on this thread just don't want to be "Using double type syntax, obviously, [it] isn't valuable."

Disabling the rule altogether allows TS to resolve FC on "arrow-function-defined" components. So that is what I and others have done -- disable this rule for the sake of TS being able to do the job itself.

Dude, I see you. You are still advocating for the rule -- you seem to have context that others on this thread don't.

SO! In the interest of becoming a better developer here -- help me understand the pitfalls where "types lie to you".

> the impact of this rule affects "observable runtime properties"

Like `mobx` and `React.Context` observed values 🤔? 

Googling around for various permutations of "observable runtime properties" specifically in the context of core React Components doesn't yield much for me, so I'd love to learn more about what you mean here (for the sake of writing better code and truly understanding where you are coming from).

I would love some examples of "arrow function(s), ~and~ causing negative runtime impact" because I am unaware of how the code example I provided is setting me up for failure in a production environment. I tried to understand this through your previous comments, but all lack links the might shed more light on the points you are concerned about.

In a way you I hope you see I am being genuine here -- help understand your concerns so that I find a way to enable this rule in a way to bring my team success.

Because otherwise, I am not going to make a decision based on your opinion (no offense). 

Data Driven Decisions are Supreme. Help me see what you see.

**Comment by ljharb** [comment]
Primarily, it's that arrow functions have an implicit/inferred name, not an explicit one, so there's a much higher likelihood than devs realize that their component won't end up with a name at all, greatly harming debuggability.

There are many other runtime differences between arrow and normal functions, but none of them really apply to a react component.

**Comment by jiminikiz** [comment]
Bro... reiterating your well established point while not at all addressing anything I asked is a weak case for following any advice you have to give 🤷

**Comment by ljharb** [comment]
You explicitly said
> I would love some examples of "arrow function(s), and causing negative runtime impact"

and that’s what i provided. please don’t call me “bro”.

**Comment by jiminikiz** [comment]
You got it. No "bro". You provided nothing. But we don't need to internet fight about it.

**Comment by linonetwo** [comment]
One use case is I expose `export type IRenderer = FC<IRendererProps>;` in my lib,

And on user side, he uses `export const` _`Pawn`_`: IRenderer = (props) => { //... ` and get this error.

**Comment by dbwcooper** [comment]
I had this problem and I solved it, although I still don't understand why it worked

```javascript
import { useEffect } from 'react';

interface Props {
  name: string;
}

const App: React.FC<Props> = ({ name }) => {
  useEffect(() => {}, []);
  return <div>{name}</div>;
};

export default App;

```
error disappears after importing `React`

```javascript
import React, { useEffect } from 'react';

interface Props {
  name: string;
}

const App: React.FC<Props> = ({ name }) => {
  useEffect(() => {}, []);
  return <div>{name}</div>;
};

export default App;

```

**Comment by ljharb** [comment]
Because React.FC doesn’t make sense if React isn’t in scope.

**Comment by ericbenedetto16** [comment]
> Because React.FC doesn’t make sense if React isn’t in scope.

@ljharb Now that an explicit React import is not required to use JSX, that shouldn't be the case for folks on React 17+ ? Needing to have `import React from 'react'` or `import type { FC } from 'react'` is a poor experience especially when the type definitions point to the same file.

**Comment by ljharb** [comment]
@ericbenedetto16 with that transform - which can be used earlier than react 17, and doesn't have to be used after - React doesn't need to be in scope for jsx syntax, is all. `React` isn't magically in-scope unless you import it in.

The quality of that experience is irrelevant; it's just how JavaScript works.

**Comment by parthcitco** [comment]
is there any solution for this ?
